zebra-chain/src/orchard/tree.rs

```diff
@@ -154,70 +153,104 @@ impl ZcashDeserialize for Root {
     }
 }
 
-/// Orchard Note Commitment Tree
+/// Orchard Incremental Note Commitment Tree
 #[derive(Clone, Debug, Default, Eq, PartialEq)]
 struct NoteCommitmentTree {
-    /// The root node of the tree (often used as an anchor).
-    root: Root,
-    /// The height of the tree (maximum height for Orchard is 32).
-    height: u8,
-    /// The number of leaves (note commitments) in this tree.
-    count: u32,
+    left: Option<pallas::Base>,
+    right: Option<pallas::Base>,
+    parents: Vec<Option<pallas::Base>>,
 }
 
-impl From<Vec<NoteCommitment>> for NoteCommitmentTree {
-    fn from(_values: Vec<NoteCommitment>) -> Self {
-        unimplemented!();
-    }
-}
-
-impl From<Vec<pallas::Base>> for NoteCommitmentTree {
-    fn from(values: Vec<pallas::Base>) -> Self {
-        if values.is_empty() {
-            return NoteCommitmentTree {
-                root: Root::default(),
-                height: 0,
-                count: 0,
-            };
+impl NoteCommitmentTree {
+    /// Adds a note commitment x-coordinate to the tree.
+    ///
+    /// The leaves of the tree are actually a base field element, the
+    /// x-coordinate of the commitment, the data that is actually stored on the
+    /// chain and input into the proof.
+    ///
+    /// Returns an error if the tree is full.
+    #[allow(clippy::identity_op)]
+    pub fn append(&mut self, cm_x: pallas::Base) -> Result<(), ()> {
+        if self.is_complete() {
+            // Tree is full
+            return Err(());
         }
 
-        let count = values.len() as u32;
-        let mut height = 0u8;
-        let mut current_layer: VecDeque<pallas::Base> = values.into_iter().collect();
-
-        while usize::from(height) < MERKLE_DEPTH {
-            let mut next_layer_up = vec![];
-
-            while !current_layer.is_empty() {
-                let left = current_layer.pop_front().unwrap();
-                let right;
-                if current_layer.is_empty() {
-                    right = EMPTY_ROOTS[height as usize];
-                } else {
-                    right = current_layer.pop_front().unwrap();
+        match (self.left, self.right) {
+            (None, _) => self.left = Some(cm_x),
+            (_, None) => self.right = Some(cm_x),
+            (Some(l), Some(r)) => {
+                let mut combined =
+                    merkle_crh_orchard((MERKLE_DEPTH - 1 - 0) as u8, Some(l), Some(r));
+                self.left = Some(cm_x);
+                self.right = None;
+
+                for i in 0..MERKLE_DEPTH {
+                    if i < self.parents.len() {
+                        if let Some(p) = self.parents[i] {
```

Comment:
Readability: this loop might be easier to read:
* as a match statement
* by combining the "`None` parent" and "extend parents" cases
* using take on the `parents` values, and on an optional `combined` value

Or those changes might make it less readable - I'm not sure.

Here's all of those changes at once.
What do you think?
```rust
                    self.parents.resize(i, None);
                    match self.parents[i].take() {
                        None => self.parents[i] = combined.take(),
                        Some(p) => combined = Some(merkle_crh_orchard(
                                (MERKLE_DEPTH - 1 - (i + 1)) as u8,
                                Some(p),
                                combined,
                            )),
                    }
                    if combined.is_none() {
                        break;
                    }
```